common/beerocks/bcl/include/bcl/network/socket.h

```diff
@@ -43,28 +43,83 @@ typedef unsigned __int64 ssize_t;
 
 class Socket {
 public:
-    explicit Socket(SOCKET s, long readTimeout = 1000);
-    Socket(SOCKET s, const std::string &peer_ip, int port_port, long readTimeout = 1000);
-    Socket(const std::string &uds_path = std::string(), long readTimeout = 1000);
+    // explicit Socket(SOCKET s, long readTimeout = 1000);
+    // Socket(SOCKET s, const std::string &peer_ip, int port_port, long readTimeout = 1000);
+    // Socket(const std::string &uds_path = std::string(), long readTimeout = 1000);
+    static Socket *socketFactory(SOCKET s, long readTimeout = 1000);
     virtual ~Socket();
 
     enum SocketMode { SocketModeBlocking, SocketModeNonBlocking };
 
-    ssize_t getBytesReady();
-    ssize_t readBytes(uint8_t *buf, size_t buf_size, bool blocking, size_t buf_len = 0,
-                      bool isPeek = false, bool waitAll = false);
+    virtual ssize_t getBytesReady()                                      = 0;
+    virtual ssize_t readBytes(uint8_t *buf, size_t buf_size, bool blocking, size_t buf_len = 0,
+                              bool isPeek = false, bool waitAll = false) = 0;
     /**
      * @brief read and discard specified number of bytes
      *
      * @return error code if an error occured or number of discarded bytes otherwise
      */
     ssize_t discardBytes(size_t bytes_to_discard);
-    ssize_t writeBytes(const uint8_t *buf, size_t buf_len, int port = 0,
-                       struct sockaddr_in addr_in = {});
-    ssize_t writeString(std::string msg)
+    virtual ssize_t writeBytes(const uint8_t *buf, size_t buf_len, int port = 0,
+                               struct sockaddr_in addr_in = {}) = 0;
+    virtual ssize_t writeString(std::string msg)
     {
         return writeBytes((const uint8_t *)msg.c_str(), (size_t)msg.length());
     }
+    virtual bool setWriteTimeout(long msec) = 0;
+    virtual bool setReadTimeout(long msec)  = 0;
+    virtual void closeSocket()              = 0;
+    virtual bool isOpen()                   = 0;
+    virtual SOCKET getSocketFd()            = 0;
```

Comment:
I personally don't believe that it's necessary to make *all* functions pure virtual to make a mockable class - but @mariomaz please correct me if I'm wrong. I think the parts which are simply "struct-like" can stay. Even the setters - you could claim that it's relevant to define an expectation for a setter, but I don't think that is correct - the expectation should be on whatever other mocked functionality is using that information.

On the other hand, this does require you to think about testing at the time you define the interface, why just making everything pure virtual is more straightforward. And it is definitely better if you don't have to think about everything at once. So maybe I'm wrong in this.

Reply:
Same answer as [here](https://github.com/prplfoundation/prplMesh/pull/1449#discussion_r443612127) 

Interfaces are contracts and IMO they should not contain anything about a particular implementation.